gnomad/variant_qc/evaluation.py

```diff
@@ -116,108 +74,78 @@ def quantiles_to_bin_boundaries(quantiles: List[int]) -> Dict:
+                hl.scan.count_where(bin_ht[f"_filter_{bin_id}"]),
             )
-            for rname in bin_stats
+            for bin_id in bin_expr
         }
     )
+    bin_ht = bin_ht.key_by("locus", "alleles")
 
+    # Annotate globals with variant counts per group defined by bin_expr. This is used to determine bin assignment
     bin_ht = bin_ht.annotate_globals(
-        bin_stats=hl.literal(
-            bin_stats,
-            dtype=hl.tstruct(
+        bin_group_variant_counts=bin_ht.aggregate(
+            hl.Struct(
                 **{
-                    bin_id: hl.tstruct(
-                        n=hl.tint64,
-                        quantiles=hl.tarray(hl.tfloat64),
-                        bin_boundaries=hl.tarray(hl.tfloat64),
-                        global_bin_indices=hl.tarray(hl.tint32),
-                        merged_bins=hl.tdict(hl.tint32, hl.tint32),
-                    )
+                    bin_id: hl.agg.filter(bin_ht[f"_filter_{bin_id}"], hl.agg.count(),)
                     for bin_id in bin_expr
                 }
-            ),
-        )
-    )
-
-    # Annotate the bin as the index in the unique boundaries array
-    bin_ht = bin_ht.annotate(
-        **{
-            bin_id: hl.or_missing(
-                bin_ht[f"_filter_{bin_id}"],
-                hl.binary_search(
-                    bin_ht.bin_stats[bin_id].bin_boundaries, bin_ht._score
-                ),
             )
-            for bin_id in bin_expr
-        }
+        )
     )
 
-    # Convert the bin to global bin by expanding merged bins, that is:
-    # If a value falls in a bin that needs expansion, assign it randomly to one of the expanded bins
-    # Otherwise, simply modify the bin to its global index (with expanded bins that is)
+    logger.info(f"Binning ranked rows into {n_bins}...")
     bin_ht = bin_ht.select(
         "snv",
         **{
-            bin_id: hl.if_else(
-                bin_ht.bin_stats[bin_id].merged_bins.contains(bin_ht[bin_id]),
-                bin_ht.bin_stats[bin_id].global_bin_indices[bin_ht[bin_id]]
-                + hl.int(
-                    hl.rand_unif(
-                        0, bin_ht.bin_stats[bin_id].merged_bins[bin_ht[bin_id]] + 1
+            bin_id: hl.int(
+                hl.floor(
+                    (
+                        n_bins
+                        * (
+                            bin_ht[f"{bin_id}_rank"]
+                            / hl.float64(bin_ht.bin_group_variant_counts[bin_id])
+                        )
                     )
-                ),
-                bin_ht.bin_stats[bin_id].global_bin_indices[bin_ht[bin_id]],
+                    + 1
+                )
             )
             for bin_id in bin_expr
         },
     )
 
     if desc:
         bin_ht = bin_ht.annotate(
-            **{bin_id: n_bins - bin_ht[bin_id] for bin_id in bin_expr}
+            **{bin_id: n_bins - bin_ht[bin_id] + 1 for bin_id in bin_expr}
```

Comment:
Note that currently, because there are 101 bins labeled (1,101) in ascending order, this code gives bins labeled (100,0) in descending order. Once the previous comment is addressed and we have 100 bins labeled (1,100) in ascending order, this code will work as intended and give bins labeled (100,1) in descending order.

Reply:
same comment as above